cwms-data-api/src/main/java/cwms/cda/data/dao/JooqDao.java

```diff
@@ -434,6 +434,47 @@ private static InvalidItemException buildInvalidItem(RuntimeException input) {
         return exception;
     }
 
+    public static boolean isInvalidUnits(RuntimeException input) {
+        boolean retVal = false;
+
+        Optional<SQLException> optional = getSqlException(input);
+        if (optional.isPresent()) {
+            SQLException sqlException = optional.get();
+            String message = sqlException.getLocalizedMessage();
+            int errorCode = sqlException.getErrorCode();
+
+            retVal = errorCode == 20998
+                    && message.contains("ORA-20102: The unit")
+                    && message.contains("is not a recognized CWMS Database unit for the")
+                ;
+        }
+        return retVal;
+    }
+
+    private static InvalidItemException buildInvalidUnits(RuntimeException input) {
+
+        Throwable cause = input;
+        if (input instanceof DataAccessException) {
+            DataAccessException dae = (DataAccessException) input;
+            cause = dae.getCause();
+        }
+
+        String localizedMessage = cause.getLocalizedMessage();
+        if (localizedMessage != null) {
+            // skip ahead in localizedMessage to "ORA-20102:"
+            int start = localizedMessage.indexOf("ORA-20102:");
+            if (start >= 0) {
```

Comment:
I say skip past the `ORA-20102:` as well.

Reply:
ok. done